### ISSUE TITLE
TFA broken after update from 1.14.x to 1.15.x.

### ISSUE THREAD

**Issue by Kusoneko** [go-gitea/gitea#16933]
- Gitea version (or commit ref): 1.15.1
- Git version: 2.33.0
- Operating system: Arch Linux
  
  Built with bindata.
  
  Running from systemd service file.
  
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
- Log gist: Not Applicable

## Description

After updating from 1.14.6 to 1.15.0 (presumably when the issue happened as I didn't try login in then and then from 1.15.0 to 1.15.1, I couldn't login with either my scratch code or my TFA TOTP generated code. I had to manually go delete my entry in my MySQL database in the two_factor table to be able to login again. Then I went to set it up again, and using both a desktop TOTP generator and copying the secret into it, as well as using the QR code for a mobile TFA app and made both generate the 6-digit codes, and both generate the exact same codes and switch codes at the exact same time, however, the Enroll into TFA page on my gitea instance is claiming that the new passcodes generated by both using the secret it provides is invalid somehow. I'm assuming there's some form of breaking change that happened with TFA in between 1.14.x and 1.15.x otherwise this wouldn't make any sense, despite my not seeing any TFA related stuff in the changelog for the breaking changes of 1.15.0. I'm essentially wanting to find out what I need to do to fix it. So far I haven't changed anything when building the new versions, only built them and then restarted gitea via `# systemctl restart gitea`. If there is any manual changes I need to do to fix this in config files or whatever, I would appreciate having them noted somewhere.

## Screenshots

Not Applicable

**Comment by zeripath** [comment]
Did you backup and restore using gitea dump?

**Comment by Kusoneko** [comment]
No, I did not. If I should have how do I fix this now? Note that besides TFA everything else appears to work fine on both the web app and using git commands locally and pushing/pulling from the gitea instance, so I might have gotten off lightly if that's something that should be done before upgrading versions like this.

**Comment by techknowlogick** [comment]
@zeripath could this be related to secret not being set?

**Comment by Kusoneko** [comment]
> @zeripath could this be related to secret not being set?

I'd doubt it personally, if we're talking about my app.ini file, I just did `# cat /etc/gitea/app.ini | grep SECRET` and I have LFS_JWT_SECRET, SECRET_KEY and JWT_SECRET set. RECAPTCHA_SECRET, HCAPTCHA_SECRET and MINIO_SECRET_ACCESS_KEY unset.

**Comment by zeripath** [comment]
> @zeripath could this be related to secret not being set?

that would explain it failing only once I think.

I'm a bit stumped. The usual failure for TOTP is something to do with time offsets but I don't understand why it would have worked previously and not now. 

There's nothing that immediately comes to mind as a possible cause - except maybe - just maybe some weird cookie issue.

I can't immediately replicate it here, and so the only thing I can think of is for you to add extra logging and build gitea.

**Comment by Kusoneko** [comment]
> I can't immediately replicate it here, and so the only thing I can think of is for you to add extra logging and build gitea.

I'm not sure what you mean by building gitea after adding extra logging, I just changed to log level to Debug in app.ini and restarted gitea, and couldn't find anywhere in the docs a build setting to use to enable further logging, but I'm missing something clearly. First, the log only contains the following for TFA stuff that I can see:

```
Sep 02 20:58:05 kusoneko.moe gitea[74645]: 2021/09/02 20:58:05 Started POST /user/settings/security/two_factor/enroll for [::1]:42432
Sep 02 20:58:05 kusoneko.moe gitea[74645]: 2021/09/02 20:58:05 ...s/context/context.go:740:1() [D] Session ID: probably_private
Sep 02 20:58:05 kusoneko.moe gitea[74645]: 2021/09/02 20:58:05 ...s/context/context.go:741:1() [D] CSRF Token: probably_private
Sep 02 20:58:05 kusoneko.moe gitea[74645]: 2021/09/02 20:58:05 Completed POST /user/settings/security/two_factor/enroll 302 Found in 21.101646ms
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 Started GET /user/settings/security/two_factor/enroll for 127.0.0.1:49340
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:740:1() [D] Session ID: probably_private
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:741:1() [D] CSRF Token: probably_private
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:185:HTML() [D] Template: user/settings/twofa_enroll
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 Completed GET /user/settings/security/two_factor/enroll 200 OK in 24.028686ms
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 Started GET /avatar/8749a188a5e167b6fee6b9613ba71d97?size=96 for [::1]:42434
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:740:1() [D] Session ID: probably_private
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:741:1() [D] CSRF Token: probably_private
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 Completed GET /avatar/8749a188a5e167b6fee6b9613ba71d97?size=96 302 Found in 2.598738ms
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 Started GET /user/events for 127.0.0.1:49342
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:740:1() [D] Session ID: probably_private
Sep 02 20:58:06 kusoneko.moe gitea[74645]: 2021/09/02 20:58:06 ...s/context/context.go:741:1() [D] CSRF Token: probably_private
```

Similarly, when looking from the dev tools of the browser, the TFA enroll page indeed starts off with a 302 followed by a 200, and when clicking the Verify button, it reloads but shows an Invalid passcode message. Nothing new seems to appear in the logs.

I did notice however that no matter how many time I leave and come back to the TFA enroll page, the TOTP secret shown on the page is ***always*** exactly the same secret (and the QR is likely the same since it gives my phone app the same TOTP secret as well), implying the problem is potentially with it not showing the right secret rather than TFA by itself not working. Hopefully that'll help a little with figuring out what's wrong.



A bit unrelated, but while we're here, I looked at the start of the logs and I'm seeing the following 2 Errors at the start:

```
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 cmd/web.go:102:runWeb() [I] Starting Gitea on PID: 74645
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...es/setting/markup.go:113:createMarkupSanitizerRule() [E] Missing required keys from markup.sanitizer.1. Must have ELEMENT and ALLOW_ATTR or ALLOW_DATA_URI_IMAGES defined!
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 cmd/web.go:146:runWeb() [I] Global init
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...es/setting/markup.go:113:createMarkupSanitizerRule() [E] Missing required keys from markup.sanitizer.1. Must have ELEMENT and ALLOW_ATTR or ALLOW_DATA_URI_IMAGES defined!
```

and the following ton of warnings right after the PING DATABASE thing:

```
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table public_key column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table repository column description db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table repository column topics db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table action column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table issue column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table pull_request column conflicted_files db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table comment column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table comment column patch db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table milestone column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table release column note db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table login_source column cfg db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table webhook Column is_system_webhook db default is 0, struct default is
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table webhook Column is_system_webhook db nullable is false, struct nullable is true
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table webhook column url db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table webhook column secret db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table webhook column events db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table webhook column meta db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table hook_task column payload_content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table hook_task column request_content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table hook_task column response_content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table notice column description db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table email_address Column lower_email db nullable is true, struct nullable is false
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table gpg_key column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table gpg_key column emails db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table gpg_key_import column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table repo_unit column config db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table external_login_user column raw_data db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table external_login_user column avatar_url db type is TEXT, struct type is VARCHAR(255)
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table external_login_user column access_token db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table external_login_user column access_token_secret db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table external_login_user column refresh_token db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column whitelist_user_i_ds db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column whitelist_team_i_ds db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column merge_whitelist_user_i_ds db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column merge_whitelist_team_i_ds db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column status_check_contexts db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column approvals_whitelist_user_i_ds db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table protected_branch column approvals_whitelist_team_i_ds db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table commit_status column target_url db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table commit_status column description db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table commit_status column context db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table lfs_lock column path db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table review column content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table oauth2_application column redirect_uris db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table task column payload_content db type is MEDIUMTEXT, struct type is TEXT
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table language_stat Column language db nullable is true, struct nullable is false
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table project column board_type db type is INT(11), struct type is INT UNSIGNED
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table project column type db type is INT(11), struct type is INT UNSIGNED
Sep 02 20:49:28 kusoneko.moe gitea[74645]: 2021/09/02 20:49:28 ...om/urfave/cli/app.go:524:HandleAction() [W] Table session has column created_unix but struct has not related field
```

Are these errors and warnings fixable?

Thank you!

**Comment by zeripath** [comment]
As per our FAQ it is safe to ignore the table warnings although `gitea doctor recreate-table` should make them go away. 

`[E] Missing required keys from markup.sanitizer.1. Must have ELEMENT and ALLOW_ATTR or ALLOW_DATA_URI_IMAGES defined!`

This implies that, like many people, you have copied app.ini.sample/app.example.ini straight across as your app.ini despite the very large warning at the top of that file NOT to do that. (We have had to comment almost the whole file out in 1.15+ because of this.)

You need to seriously simplify your app.ini dropping whole sections from it that you have not explicitly set. My own app.ini is 75 lines long and really that is still too long. 

---

These app.ini issues aren't obviously the cause of the problem (except where maybe you've incorrectly overconfigured your cookies?)

---

Looking at your logs I see that the IP address you are posting from to appears to be flipping from 127.0.0.1 to [::1] which is concerning that you might be associated with different sessions etc.

You've sanitized the session id away so I can't check that theyre the same and you've not given me the logs from the GET before the enroll POST so I can't check that either. We need to see a little more of the interaction before the POST.

It's helpful to double check if the issue is a browser specific one - i.e. are there multiple weird cookies on your browser that need to be erased.

You haven't told us whether there's a reverse proxy involved or if Gitea is mounted on a suburl. Are you definitely accessing gitea by its ROOT_URL - maybe the cookies aren't being passed down - the session id being the same should assert that.

---

Regarding additional logging:

You asserted that debug logs were not helpful - although it does appear that they may help shed further light on this issue so follow the instructions given in issue proforma and give us those logs. My suggestion about us adding extra logging was that we would need to compile in some extra logging at certain places. 

Give us the logs as per https://docs.gitea.io/en-us/logging-configuration/#debugging-problems sanitising only the minimum and asserting where things are the same if you've sanitized them away.

Explain your setup, e.g. reverse proxy or not, how are you running gitea, what is your systemd unit file, what version of mysql/mariadb.

Show us your app.ini - you can send me the unsanitised one on discord if you message me directly.

Check if on private browsing or different browser if things work. The problem may be cookie related.

**Comment by Kusoneko** [comment]
> These app.ini issues aren't obviously the cause of the problem (except where maybe you've incorrectly overconfigured your cookies?)

Thanks, that fixed the errors and most of the warnings, I assume the last 2 that gitea doctor didn't fix, where it says the table's field should be int unsigned and that it's currently int(10) unsigned is far from the end of the world.

Well, I did configure them a bit, but I just swapped it from storing the sessions from file mode to memory mode, and deleted all my cookies for the instance to see if something changed, indeed, the TOTP secret has changed afterwards, but it's sticking to the new secret now, and still refusing codes. This is a really odd issue.
 
> Looking at your logs I see that the IP address you are posting from to appears to be flipping from 127.0.0.1 to [::1] which is concerning that you might be associated with different sessions etc.
>
> You've sanitized the session id away so I can't check that they're the same and you've not given me the logs from the GET before the enroll POST so I can't check that either. We need to see a little more of the interaction before the POST.
> 
> It's helpful to double check if the issue is a browser specific one - i.e. are there multiple weird cookies on your browser that need to be erased.

They were identical for that specific snippet.

> You haven't told us whether there's a reverse proxy involved or if Gitea is mounted on a suburl. Are you definitely accessing gitea by its ROOT_URL - maybe the cookies aren't being passed down - the session id being the same should assert that.
> 
> Regarding additional logging:
> 
> You asserted that debug logs were not helpful - although it does appear that they may help shed further light on this issue so follow the instructions given in issue proforma and give us those logs. My suggestion about us adding extra logging was that we would need to compile in some extra logging at certain places.
> 
> Give us the logs as per https://docs.gitea.io/en-us/logging-configuration/#debugging-problems sanitising only the minimum and asserting where things are the same if you've sanitized them away.
> 
> Explain your setup, e.g. reverse proxy or not, how are you running gitea, what is your systemd unit file, what version of mysql/mariadb.
> 
> Show us your app.ini - you can send me the unsanitised one on discord if you message me directly.
> 
> Check if on private browsing or different browser if things work. The problem may be cookie related.

Right, my gitea instance is configured to listen to localhost:3001 (unless I'm misunderstanding the app.ini comments), except for ssh requests which is actually using the git user and openssh's sshd. SSH works fine, so that's not the issue. The server is actually serving multiple (mostly personal) web apps and services, so to prevent port collision issues and all that, all of them are being run behind nginx as a reverse proxy (except for one or 2 webpages that nginx serves itself), unless they're not web apps and therefore don't require port 443/80 to be used without fiddling around with ports in a browser.

The results of `$ mysql --version`: `mysql  Ver 15.1 Distrib 10.6.4-MariaDB, for Linux (x86_64) using readline 5.1`

SystemD service file:

```
[Unit]
Description=Gitea (Git with a cup of tea)
After=syslog.target
After=network.target
Requires=mariadb.service

[Service]
# Modify these two values and uncomment them if you have
# repos with lots of files and get an HTTP error 500 because
# of that
###
#LimitMEMLOCK=infinity
#LimitNOFILE=65535
RestartSec=2s
Type=simple
User=git
Group=git
WorkingDirectory=/var/lib/gitea/
ExecStart=/usr/local/bin/gitea web --config /etc/gitea/app.ini --custom-path /var/lib/gitea/custom --work-path /var/lib/gitea --port 3001
Restart=always
Environment=USER=git HOME=/home/git GITEA_WORK_DIR=/var/lib/gitea

[Install]
WantedBy=multi-user.target
```

The logs after stopping gitea, removing all my cookies from my browser, then restarting gitea, while I go speedrun to the TFA enroll page are in the attached file. The Session IDs were renamed id# and the CSRF Tokens token# with sed, so they should have the same number for all occurrences of each.

The app.ini configuration for the looks like this:

```
; App name that shows in every page title
APP_NAME = Gitea: Git with a cup of tea
; Change it if you run locally
RUN_USER = git
; Application run mode, affects performance and debugging. Either "dev", "prod" or "test", default is "prod"
RUN_MODE = prod

[project]
; Default templates for project boards
PROJECT_BOARD_BASIC_KANBAN_TYPE = To Do, In Progress, Done
PROJECT_BOARD_BUG_TRIAGE_TYPE   = Needs Triage, High Priority, Low Priority, Closed

[repository]
; Root path for storing all repository data. It must be an absolute path. By default it is stored in a sub-directory of `APP_DATA_PATH`.
ROOT                                           = /home/git/gitea-repositories
; The script type this server supports. Usually this is `bash`, but some users report that only `sh` is available.
SCRIPT_TYPE                                    = bash
; DETECTED_CHARSETS_ORDER tie-break order for detected charsets.
; If the charsets have equal confidence, tie-breaking will be done by order in this list
; with charsets earlier in the list chosen in preference to those later.
; Adding "defaults" will place the unused charsets at that position.
DETECTED_CHARSETS_ORDER                        = UTF-8, UTF-16BE, UTF-16LE, UTF-32BE, UTF-32LE, ISO-8859, windows-1252, ISO-8859, windows-1250, ISO-8859, ISO-8859, ISO-8859, windows-1253, ISO-8859, windows-1255, ISO-8859, windows-1251, windows-1256, KOI8-R, ISO-8859, windows-1254, Shift_JIS, GB18030, EUC-JP, EUC-KR, Big5, ISO-2022, ISO-2022, ISO-2022, IBM424_rtl, IBM424_ltr, IBM420_rtl, IBM420_ltr
; Default ANSI charset to override non-UTF-8 charsets to
ANSI_CHARSET                                   =
; Force every new repository to be private
FORCE_PRIVATE                                  = false
; Default privacy setting when creating a new repository, allowed values: last, private, public. Default is last which means the last setting used.
DEFAULT_PRIVATE                                = last
; Default private when using push-to-create
DEFAULT_PUSH_CREATE_PRIVATE                    = true
; Global limit of repositories per user, applied at creation time. -1 means no limit
MAX_CREATION_LIMIT                             = -1
; Mirror sync queue length, increase if mirror syncing starts hanging
MIRROR_QUEUE_LENGTH                            = 1000
; Patch test queue length, increase if pull request patch testing starts hanging
PULL_REQUEST_QUEUE_LENGTH                      = 1000
; Preferred Licenses to place at the top of the List
; The name here must match the filename in conf/license or custom/conf/license
PREFERRED_LICENSES                             = Apache License 2.0,MIT License
; Disable the ability to interact with repositories using the HTTP protocol
DISABLE_HTTP_GIT                               = false
; Value for Access-Control-Allow-Origin header, default is not to present
; WARNING: This maybe harmful to you website if you do not give it a right value.
ACCESS_CONTROL_ALLOW_ORIGIN                    =
; Force ssh:// clone url instead of scp-style uri when default SSH port is used
USE_COMPAT_SSH_URI                             = false
; Close issues as long as a commit on any branch marks it as fixed
DEFAULT_CLOSE_ISSUES_VIA_COMMITS_IN_ANY_BRANCH = false
; Allow users to push local repositories to Gitea and have them automatically created for a user or an org
ENABLE_PUSH_CREATE_USER                        = true
ENABLE_PUSH_CREATE_ORG                         = true
; Comma separated list of globally disabled repo units. Allowed values: repo.issues, repo.ext_issues, repo.pulls, repo.wiki, repo.ext_wiki
DISABLED_REPO_UNITS                            =
; Comma separated list of default repo units. Allowed values: repo.code, repo.releases, repo.issues, repo.pulls, repo.wiki, repo.projects.
; Note: Code and Releases can currently not be deactivated. If you specify default repo units you should still list them for future compatibility.
; External wiki and issue tracker can't be enabled by default as it requires additional settings.
; Disabled repo units will not be added to new repositories regardless if it is in the default list.
DEFAULT_REPO_UNITS                             = repo.code,repo.releases,repo.issues,repo.pulls,repo.wiki,repo.projects
; Prefix archive files by placing them in a directory named after the repository
PREFIX_ARCHIVE_FILES                           = true
; Disable the creation of new mirrors. Pre-existing mirrors remain valid.
DISABLE_MIRRORS                                = false
; Disable migrating feature.
DISABLE_MIGRATIONS                             = false
; The default branch name of new repositories
DEFAULT_BRANCH                                 = master
; Allow adoption of unadopted repositories
ALLOW_ADOPTION_OF_UNADOPTED_REPOSITORIES       = false
; Allow deletion of unadopted repositories
ALLOW_DELETION_OF_UNADOPTED_REPOSITORIES       = false

[repository.editor]
; List of file extensions for which lines should be wrapped in the Monaco editor
; Separate extensions with a comma. To line wrap files without an extension, just put a comma
LINE_WRAP_EXTENSIONS   = .txt,.md,.markdown,.mdown,.mkd,
; Valid file modes that have a preview API associated with them, such as api/v1/markdown
; Separate the values by commas. The preview tab in edit mode won't be displayed if the file extension doesn't match
PREVIEWABLE_FILE_MODES = markdown

[repository.local]
; Path for local repository copy. Defaults to `tmp/local-repo`
LOCAL_COPY_PATH = tmp/local-repo

[repository.upload]
; Whether repository file uploads are enabled. Defaults to `true`
ENABLED       = true
; Path for uploads. Defaults to `data/tmp/uploads` (tmp gets deleted on gitea restart)
TEMP_PATH     = data/tmp/uploads
; Comma-separated list of allowed file extensions (`.zip`), mime types (`text/plain`) or wildcard type (`image/*`, `audio/*`, `video/*`). Empty value or `*/*` allows all types.
ALLOWED_TYPES =
; Max size of each file in megabytes. Defaults to 3MB
FILE_MAX_SIZE = 3
; Max number of files per upload. Defaults to 5
MAX_FILES     = 5

[repository.pull-request]
; List of prefixes used in Pull Request title to mark them as Work In Progress
WORK_IN_PROGRESS_PREFIXES                     = WIP:,[WIP]
; List of keywords used in Pull Request comments to automatically close a related issue
CLOSE_KEYWORDS                                = close,closes,closed,fix,fixes,fixed,resolve,resolves,resolved
; List of keywords used in Pull Request comments to automatically reopen a related issue
REOPEN_KEYWORDS                               = reopen,reopens,reopened
; In the default merge message for squash commits include at most this many commits
DEFAULT_MERGE_MESSAGE_COMMITS_LIMIT           = 50
; In the default merge message for squash commits limit the size of the commit messages to this
DEFAULT_MERGE_MESSAGE_SIZE                    = 5120
; In the default merge message for squash commits walk all commits to include all authors in the Co-authored-by otherwise just use those in the limited list
DEFAULT_MERGE_MESSAGE_ALL_AUTHORS             = false
; In default merge messages limit the number of approvers listed as Reviewed-by: to this many
DEFAULT_MERGE_MESSAGE_MAX_APPROVERS           = 10
; In default merge messages only include approvers who are official
DEFAULT_MERGE_MESSAGE_OFFICIAL_APPROVERS_ONLY = true

[repository.issue]
; List of reasons why a Pull Request or Issue can be locked
LOCK_REASONS = Too heated,Off-topic,Resolved,Spam

[repository.release]
; Comma-separated list of allowed file extensions (`.zip`), mime types (`text/plain`) or wildcard type (`image/*`, `audio/*`, `video/*`). Empty value or `*/*` allows all types.
ALLOWED_TYPES =

[repository.signing]
; GPG key to use to sign commits, Defaults to the default - that is the value of git config --get user.signingkey
; run in the context of the RUN_USER
; Switch to none to stop signing completely
SIGNING_KEY         = default
; If a SIGNING_KEY ID is provided and is not set to default, use the provided Name and Email address as the signer.
; These should match a publicized name and email address for the key. (When SIGNING_KEY is default these are set to
; the results of git config --get user.name and git config --get user.email respectively and can only be overrided
; by setting the SIGNING_KEY ID to the correct ID.)
SIGNING_NAME        =
SIGNING_EMAIL       =
; Sets the default trust model for repositories. Options are: collaborator, committer, collaboratorcommitter
DEFAULT_TRUST_MODEL = collaborator
; Determines when gitea should sign the initial commit when creating a repository
; Either:
; - never
; - pubkey: only sign if the user has a pubkey
; - twofa: only sign if the user has logged in with twofa
; - always
; options other than none and always can be combined as comma separated list
INITIAL_COMMIT      = always
; Determines when to sign for CRUD actions
; - as above
; - parentsigned: requires that the parent commit is signed.
CRUD_ACTIONS        = pubkey, twofa, parentsigned
; Determines when to sign Wiki commits
; - as above
WIKI                = never
; Determines when to sign on merges
; - basesigned: require that the parent of commit on the base repo is signed.
; - commitssigned: require that all the commits in the head branch are signed.
; - approved: only sign when merging an approved pr to a protected branch
MERGES              = pubkey, twofa, basesigned, commitssigned

[cors]
; More information about CORS can be found here: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#The_HTTP_response_headers
; enable cors headers (disabled by default)
ENABLED           = false
; scheme of allowed requests
SCHEME            = http
; list of requesting domains that are allowed
ALLOW_DOMAIN      = *
; allow subdomains of headers listed above to request
ALLOW_SUBDOMAIN   = false
; list of methods allowed to request
METHODS           = GET,HEAD,POST,PUT,PATCH,DELETE,OPTIONS
; max time to cache response
MAX_AGE           = 10m
; allow request with credentials
ALLOW_CREDENTIALS = false

[ui]
; Number of repositories that are displayed on one explore page
EXPLORE_PAGING_NUM      = 20
; Number of issues that are displayed on one page
ISSUE_PAGING_NUM        = 10
; Number of maximum commits displayed in one activity feed
FEED_MAX_COMMIT_NUM     = 5
; Number of items that are displayed in home feed
FEED_PAGING_NUM         = 20
; Number of maximum commits displayed in commit graph.
GRAPH_MAX_COMMIT_NUM    = 100
; Number of line of codes shown for a code comment
CODE_COMMENT_LINES      = 4
; Value of `theme-color` meta tag, used by Android >= 5.0
; An invalid color like "none" or "disable" will have the default style
; More info: https://developers.google.com/web/updates/2014/11/Support-for-theme-color-in-Chrome-39-for-Android
THEME_COLOR_META_TAG    = `#6cc644`
; Max size of files to be displayed (default is 8MiB)
MAX_DISPLAY_FILE_SIZE   = 8388608
; Whether the email of the user should be shown in the Explore Users page
SHOW_USER_EMAIL         = true
; Set the default theme for the Gitea install
DEFAULT_THEME           = gitea
; All available themes. Allow users select personalized themes regardless of the value of `DEFAULT_THEME`.
THEMES                  = gitea,arc-green
; All available reactions users can choose on issues/prs and comments.
; Values can be emoji alias (:smile:) or a unicode emoji.
; For custom reactions, add a tightly cropped square image to public/emoji/img/reaction_name.png
REACTIONS               = +1, -1, laugh, hooray, confused, heart, rocket, eyes
; Whether the full name of the users should be shown where possible. If the full name isn't set, the username will be used.
DEFAULT_SHOW_FULL_NAME  = false
; Whether to search within description at repository search on explore page.
SEARCH_REPO_DESCRIPTION = true
; Whether to enable a Service Worker to cache frontend assets
USE_SERVICE_WORKER      = true

[ui.admin]
; Number of users that are displayed on one page
USER_PAGING_NUM   = 50
; Number of repos that are displayed on one page
REPO_PAGING_NUM   = 50
; Number of notices that are displayed on one page
NOTICE_PAGING_NUM = 25
; Number of organizations that are displayed on one page
ORG_PAGING_NUM    = 50

[ui.user]
; Number of repos that are displayed on one page
REPO_PAGING_NUM = 15

[ui.meta]
AUTHOR      = Gitea - Git with a cup of tea
DESCRIPTION = Gitea (Git with a cup of tea) is a painless self-hosted Git service written in Go
KEYWORDS    = go,git,self-hosted,gitea

[ui.notification]
; Control how often the notification endpoint is polled to update the notification
; The timeout will increase to MAX_TIMEOUT in TIMEOUT_STEPs if the notification count is unchanged
; Set MIN_TIMEOUT to 0 to turn off
MIN_TIMEOUT              = 10s
MAX_TIMEOUT              = 60s
TIMEOUT_STEP             = 10s
; This setting determines how often the db is queried to get the latest notification counts.
; If the browser client supports EventSource and SharedWorker, a SharedWorker will be used in preference to polling notification. Set to -1 to disable the EventSource
EVENT_SOURCE_UPDATE_TIME = 10s

[ui.svg]
; Whether to render SVG files as images.  If SVG rendering is disabled, SVG files are displayed as text and cannot be embedded in markdown files as images.
ENABLE_RENDER = true

[markdown]
; Render soft line breaks as hard line breaks, which means a single newline character between
; paragraphs will cause a line break and adding trailing whitespace to paragraphs is not
; necessary to force a line break.
; Render soft line breaks as hard line breaks for comments
ENABLE_HARD_LINE_BREAK_IN_COMMENTS  = true
; Render soft line breaks as hard line breaks for markdown documents
ENABLE_HARD_LINE_BREAK_IN_DOCUMENTS = false
; Comma separated list of custom URL-Schemes that are allowed as links when rendering Markdown
; for example git,magnet,ftp (more at https://en.wikipedia.org/wiki/List_of_URI_schemes)
; URLs starting with http and https are always displayed, whatever is put in this entry.
CUSTOM_URL_SCHEMES                  = git,magnet,ftp,sftp,ssh
; List of file extensions that should be rendered/edited as Markdown
; Separate the extensions with a comma. To render files without any extension as markdown, just put a comma
FILE_EXTENSIONS                     = .md,.markdown,.mdown,.mkd

[server]
; The protocol the server listens on. One of 'http', 'https', 'unix' or 'fcgi'.
PROTOCOL                              = http
DOMAIN                                = localhost
ROOT_URL                              = https://git.kusoneko.moe/
; The address to listen on. Either a IPv4/IPv6 address or the path to a unix socket.
HTTP_ADDR                             = 0.0.0.0
; The port to listen on. Leave empty when using a unix socket.
HTTP_PORT                             = 3001
; Local (DMZ) URL for Gitea workers (such as SSH update) accessing web service.
; In most cases you do not need to change the default value.
; Alter it only if your SSH server node is not the same as HTTP node.
; Do not set this variable if PROTOCOL is set to 'unix'.
LOCAL_ROOT_URL                        = http://localhost:3001/
; Domain name to be exposed in clone URL
SSH_DOMAIN                            = kusoneko.moe
; Port number to be exposed in clone URL
SSH_PORT                              = 22
; Root path of SSH directory, default is '~/.ssh', but you have to use '/home/git/.ssh'.
SSH_ROOT_PATH                         = /home/git/.ssh
; Gitea will create a authorized_keys file by default when it is not using the internal ssh server
; If you intend to use the AuthorizedKeysCommand functionality then you should turn this off.
SSH_CREATE_AUTHORIZED_KEYS_FILE       = true
; Gitea will create a authorized_principals file by default when it is not using the internal ssh server
; If you intend to use the AuthorizedPrincipalsCommand functionality then you should turn this off.
SSH_CREATE_AUTHORIZED_PRINCIPALS_FILE = true
; Determines which principals to allow
; - empty: if SSH_TRUSTED_USER_CA_KEYS is empty this will default to off, otherwise will default to email, username.
; - off: Do not allow authorized principals
; - email: the principal must match the user's email
; - username: the principal must match the user's username
; - anything: there will be no checking on the content of the principal
SSH_AUTHORIZED_PRINCIPALS_ALLOW       = email, username
; Indicate whether to check minimum key size with corresponding type
MINIMUM_KEY_SIZE_CHECK                = false
; Disable CDN even in "prod" mode
OFFLINE_MODE                          = false
DISABLE_ROUTER_LOG                    = false
; Default path for App data
APP_DATA_PATH                         = data
; Enable gzip compression for runtime-generated content, static resources excluded
ENABLE_GZIP                           = false
; Landing page, can be "home", "explore", "organizations" or "login"
; The "login" choice is not a security measure but just a UI flow change, use REQUIRE_SIGNIN_VIEW to force users to log in.
LANDING_PAGE                          = home
; Enables git-lfs support. true or false, default is false.
LFS_START_SERVER                      = true
; Where your lfs files reside, default is data/lfs.
LFS_CONTENT_PATH                      = /var/lib/gitea/data/lfs
; LFS authentication secret, change this yourself
LFS_JWT_SECRET                        = xxxx
; LFS authentication validity period (in time.Duration), pushes taking longer than this may fail.
LFS_HTTP_AUTH_EXPIRY                  = 20m
; Maximum allowed LFS file size in bytes (Set to 0 for no limit).
LFS_MAX_FILE_SIZE                     = 0
; Maximum number of locks returned per page
LFS_LOCKS_PAGING_NUM                  = 50
; Allow graceful restarts using SIGHUP to fork
ALLOW_GRACEFUL_RESTARTS               = true
; After a restart the parent will finish ongoing requests before
; shutting down. Force shutdown if this process takes longer than this delay.
; set to a negative value to disable
GRACEFUL_HAMMER_TIME                  = 60s

; Define allowed algorithms and their minimum key length (use -1 to disable a type)
[ssh.minimum_key_sizes]
ED25519 = 256
ECDSA   = 256
RSA     = 2048
; set to 1024 to switch on
DSA     = -1

[database]
; Database to use. Either "mysql", "postgres", "mssql" or "sqlite3".
DB_TYPE             = mysql
HOST                = 127.0.0.1:3306
NAME                = gitea
USER                = git
; Use PASSWD = `your password` for quoting if you use special characters in the password.
PASSWD              = xxxxx
; For MySQL, either "false" (default), "true", or "skip-verify"
SSL_MODE            = false
; For MySQL only, either "utf8" or "utf8mb4", default is "utf8mb4".
; NOTICE: for "utf8mb4" you must use MySQL InnoDB > 5.6. Gitea is unable to check this.
CHARSET             = utf8mb4
; For iterate buffer, default is 50
ITERATE_BUFFER_SIZE = 50
; Show the database generated SQL
LOG_SQL             = false
; Maximum number of DB Connect retries
DB_RETRIES          = 10
; Backoff time per DB retry (time.Duration)
DB_RETRY_BACKOFF    = 3s
; Max idle database connections on connnection pool, default is 2
MAX_IDLE_CONNS      = 2
; Database connection max life time, default is 0 or 3s mysql (See #6804 & #7071 for reasoning)
CONN_MAX_LIFETIME   = 3s
; Database maximum number of open connections, default is 0 meaning no maximum
MAX_OPEN_CONNS      = 0

[indexer]
; Issue indexer type, currently support: bleve, db or elasticsearch, default is bleve
ISSUE_INDEXER_TYPE               = bleve
; Issue indexer storage path, available when ISSUE_INDEXER_TYPE is bleve
ISSUE_INDEXER_PATH               = indexers/issues.bleve
; Issue indexer queue, currently support: channel, levelqueue or redis, default is levelqueue
ISSUE_INDEXER_QUEUE_TYPE         = levelqueue
; When ISSUE_INDEXER_QUEUE_TYPE is levelqueue, this will be the path where the queue will be saved.
; This can be overriden by `ISSUE_INDEXER_QUEUE_CONN_STR`.
; default is indexers/issues.queue
ISSUE_INDEXER_QUEUE_DIR          = indexers/issues.queue
; Batch queue number, default is 20
ISSUE_INDEXER_QUEUE_BATCH_NUMBER = 20
; Timeout the indexer if it takes longer than this to start.
; Set to zero to disable timeout.
STARTUP_TIMEOUT                  = 30s
; repo indexer by default disabled, since it uses a lot of disk space
REPO_INDEXER_ENABLED             = false
; Code search engine type, could be `bleve` or `elasticsearch`.
REPO_INDEXER_TYPE                = bleve
; Index file used for code search.
REPO_INDEXER_PATH                = indexers/repos.bleve
UPDATE_BUFFER_LEN                = 20
MAX_FILE_SIZE                    = 1048576

[queue]
; Specific queues can be individually configured with [queue.name]. [queue] provides defaults
;
; General queue queue type, currently support: persistable-channel, channel, level, redis, dummy
; default to persistable-channel
TYPE              = persistable-channel
; data-dir for storing persistable queues and level queues, individual queues will be named by their type
DATADIR           = queues/
; Default queue length before a channel queue will block
LENGTH            = 20
; Batch size to send for batched queues
BATCH_LENGTH      = 20
; Provides the suffix of the default redis/disk queue name - specific queues can be overriden within in their [queue.name] sections.
QUEUE_NAME        = _queue
; Provides the suffix of the default redis/disk unique queue set name - specific queues can be overriden within in their [queue.name] sections.
SET_NAME          = _unique
; If the queue cannot be created at startup - level queues may need a timeout at startup - wrap the queue:
WRAP_IF_NECESSARY = true
; Attempt to create the wrapped queue at max
MAX_ATTEMPTS      = 10
; Timeout queue creation
TIMEOUT           = 15m30s
; Create a pool with this many workers
WORKERS           = 1
; Dynamically scale the worker pool to at this many workers
MAX_WORKERS       = 10
; Add boost workers when the queue blocks for BLOCK_TIMEOUT
BLOCK_TIMEOUT     = 1s
; Remove the boost workers after BOOST_TIMEOUT
BOOST_TIMEOUT     = 5m
; During a boost add BOOST_WORKERS
BOOST_WORKERS     = 5

[admin]
; Disallow regular (non-admin) users from creating organizations.
DISABLE_REGULAR_ORG_CREATION = false
; Default configuration for email notifications for users (user configurable). Options: enabled, onmention, disabled
DEFAULT_EMAIL_NOTIFICATIONS  = onmention

[security]
; Whether the installer is disabled
INSTALL_LOCK                             = true
; !!CHANGE THIS TO KEEP YOUR USER DATA SAFE!!
SECRET_KEY                               = xxxxxxxxxxxxxx
; How long to remember that a user is logged in before requiring relogin (in days)
LOGIN_REMEMBER_DAYS                      = 7
COOKIE_USERNAME                          = gu
COOKIE_REMEMBER_NAME                     = gr
; Reverse proxy authentication header name of user name
REVERSE_PROXY_AUTHENTICATION_USER        = X-WEBAUTH-USER
REVERSE_PROXY_AUTHENTICATION_EMAIL       = X-WEBAUTH-EMAIL
; Interpret X-Forwarded-For header or the X-Real-IP header and set this as the remote IP for the request
REVERSE_PROXY_LIMIT                      = 1
; List of IP addresses and networks seperated by comma of trusted proxy servers. Use `*` to trust all.
REVERSE_PROXY_TRUSTED_PROXIES            = 127.0.0.0/8,::1/128
; The minimum password length for new Users
MIN_PASSWORD_LENGTH                      = 10
; Set to true to allow users to import local server paths
IMPORT_LOCAL_PATHS                       = false
; Set to false to allow users with git hook privileges to create custom git hooks.
; Custom git hooks can be used to perform arbitrary code execution on the host operating system.
; This enables the users to access and modify this config file and the Gitea database and interrupt the Gitea service.
; By modifying the Gitea database, users can gain Gitea administrator privileges.
; It also enables them to access other resources available to the user on the operating system that is running the Gitea instance and perform arbitrary actions in the name of the Gitea OS user.
; WARNING: This maybe harmful to you website or your operating system.
DISABLE_GIT_HOOKS                        = true
; Set to true to disable webhooks feature.
DISABLE_WEBHOOKS                         = false
; Set to false to allow pushes to gitea repositories despite having an incomplete environment - NOT RECOMMENDED
ONLY_ALLOW_PUSH_IF_GITEA_ENVIRONMENT_SET = true
; Comma separated list of character classes required to pass minimum complexity.
; If left empty or no valid values are specified, the default is off (no checking)
; Classes include "lower,upper,digit,spec"
PASSWORD_COMPLEXITY                      = spec
; Password Hash algorithm, either "argon2", "pbkdf2", "scrypt" or "bcrypt"
PASSWORD_HASH_ALGO                       = pbkdf2
; Set false to allow JavaScript to read CSRF cookie
CSRF_COOKIE_HTTP_ONLY                    = true
; Validate against https://haveibeenpwned.com/Passwords to see if a password has been exposed
PASSWORD_CHECK_PWN                       = true
INTERNAL_TOKEN                           = xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

[openid]
;
; OpenID is an open, standard and decentralized authentication protocol.
; Your identity is the address of a webpage you provide, which describes
; how to prove you are in control of that page.
;
; For more info: https://en.wikipedia.org/wiki/OpenID
;
; Current implementation supports OpenID-2.0
;
; Tested to work providers at the time of writing:
; - Any GNUSocial node (your.hostname.tld/username)
; - Any SimpleID provider (http://simpleid.koinic.net)
; - http://openid.org.cn/
; - openid.stackexchange.com
; - login.launchpad.net
; - <username>.livejournal.com
;
; Whether to allow signin in via OpenID
ENABLE_OPENID_SIGNIN = true
; Whether to allow registering via OpenID
; Do not include to rely on rhw DISABLE_REGISTRATION setting
ENABLE_OPENID_SIGNUP = false

[service]
; Time limit to confirm account/email registration
ACTIVE_CODE_LIVE_MINUTES                      = 180
; Time limit to perform the reset of a forgotten password
RESET_PASSWD_CODE_LIVE_MINUTES                = 180
; Whether a new user needs to confirm their email when registering.
REGISTER_EMAIL_CONFIRM                        = true
; Disallow registration, only allow admins to create accounts.
DISABLE_REGISTRATION                          = false
; Allow registration only using third-party services, it works only when DISABLE_REGISTRATION is false
ALLOW_ONLY_EXTERNAL_REGISTRATION              = false
; User must sign in to view anything.
REQUIRE_SIGNIN_VIEW                           = false
; Mail notification
ENABLE_NOTIFY_MAIL                            = true
; This setting enables gitea to be signed in with HTTP BASIC Authentication using the user's password
; If you set this to false you will not be able to access the tokens endpoints on the API with your password
; Please note that setting this to false will not disable OAuth Basic or Basic authentication using a token
ENABLE_BASIC_AUTHENTICATION                   = true
; More detail: https://github.com/gogits/gogs/issues/165
ENABLE_REVERSE_PROXY_AUTHENTICATION           = false
ENABLE_REVERSE_PROXY_AUTO_REGISTRATION        = false
ENABLE_REVERSE_PROXY_EMAIL                    = false
; Enable captcha validation for registration
ENABLE_CAPTCHA                                = false
; Default value for KeepEmailPrivate
; Each new user will get the value of this setting copied into their profile
DEFAULT_KEEP_EMAIL_PRIVATE                    = false
; Default value for AllowCreateOrganization
; Every new user will have rights set to create organizations depending on this setting
DEFAULT_ALLOW_CREATE_ORGANIZATION             = true
; Default value for DefaultOrgMemberVisible
; True will make the membership of the users visible when added to the organisation
DEFAULT_ORG_MEMBER_VISIBLE                    = false
; Default value for EnableDependencies
; Repositories will use dependencies by default depending on this setting
DEFAULT_ENABLE_DEPENDENCIES                   = true
; Dependencies can be added from any repository where the user is granted access or only from the current repository depending on this setting.
ALLOW_CROSS_REPOSITORY_DEPENDENCIES           = true
; Enable heatmap on users profiles.
ENABLE_USER_HEATMAP                           = true
; Enable Timetracking
ENABLE_TIMETRACKING                           = true
; Default value for EnableTimetracking
; Repositories will use timetracking by default depending on this setting
DEFAULT_ENABLE_TIMETRACKING                   = true
; Default value for AllowOnlyContributorsToTrackTime
; Only users with write permissions can track time if this is true
DEFAULT_ALLOW_ONLY_CONTRIBUTORS_TO_TRACK_TIME = true
; Show Registration button
SHOW_REGISTRATION_BUTTON                      = true
; Show milestones dashboard page - a view of all the user's milestones
SHOW_MILESTONES_DASHBOARD_PAGE                = true
; Default value for AutoWatchNewRepos
; When adding a repo to a team or creating a new repo all team members will watch the
; repo automatically if enabled
AUTO_WATCH_NEW_REPOS                          = true
; Default value for AutoWatchOnChanges
; Make the user watch a repository When they commit for the first time
AUTO_WATCH_ON_CHANGES                         = false
; Minimum amount of time a user must exist before comments are kept when the user is deleted.
USER_DELETE_WITH_COMMENTS_MAX_TIME            = 0

[webhook]
; Hook task queue length, increase if webhook shooting starts hanging
QUEUE_LENGTH    = 1000
; Deliver timeout in seconds
DELIVER_TIMEOUT = 5
; Allow insecure certification
SKIP_TLS_VERIFY = false
; Number of history information in each page
PAGING_NUM      = 10

[mailer]
ENABLED            = true
; Buffer length of channel, keep it as it is if you don't know what it is.
SEND_BUFFER_LEN    = 100
; Prefix displayed before subject in mail
SUBJECT_PREFIX     = [kusoneko.moe]
; Mail server
; Gmail: smtp.gmail.com:587
; QQ: smtp.qq.com:465
; Using STARTTLS on port 587 is recommended per RFC 6409.
; Note, if the port ends with "465", SMTPS will be used.
HOST               = smtppro.zoho.com:465
; Whether or not to skip verification of certificates; `true` to disable verification. This option is unsafe. Consider adding the certificate to the system trust store instead.
SKIP_VERIFY        = false
; Use client certificate
USE_CERTIFICATE    = false
; Should SMTP connect with TLS, (if port ends with 465 TLS will always be used.)
; If this is false but STARTTLS is supported the connection will be upgraded to TLS opportunistically.
IS_TLS_ENABLED     = true
; Mail from address, RFC 5322. This can be just an email address, or the `"Name" <email@example.com>` format
FROM               = noreply@kusoneko.moe
; Mailer user name and password
; Please Note: Authentication is only supported when the SMTP server communication is encrypted with TLS (this can be via STARTTLS) or `HOST=localhost`.
USER               = noreply@kusoneko.moe
; Use PASSWD = `your password` for quoting if you use special characters in the password.
PASSWD             = xxxxxxxxxxxxxxxxxxxxxxxxxxx
; Send mails as plain text
SEND_AS_PLAIN_TEXT = false
; Set Mailer Type (either SMTP, sendmail or dummy to just send to the log)
MAILER_TYPE        = smtp

[cache]
; if the cache enabled
ENABLED  = true
; Either "memory", "redis", or "memcache", default is "memory"
ADAPTER  = memory
; For "memory" only, GC interval in seconds, default is 60
INTERVAL = 60
; Time to keep items in cache if not used, default is 16 hours.
; Setting it to 0 disables caching
ITEM_TTL = 16h

; Last commit cache
[cache.last_commit]
; if the cache enabled
ENABLED       = true
; Time to keep items in cache if not used, default is 8760 hours.
; Setting it to 0 disables caching
ITEM_TTL      = 8760h
; Only enable the cache when repository's commits count great than
COMMITS_COUNT = 1000

[session]
; Either "memory", "file", or "redis", default is "memory"
PROVIDER          = memory
; Session cookie name
COOKIE_NAME       = gs
; If you use session in https only, default is false
COOKIE_SECURE     = true
; Session GC time interval in seconds, default is 86400 (1 day)
GC_INTERVAL_TIME  = 86400
; Session life time in seconds, default is 86400 (1 day)
SESSION_LIFE_TIME = 86400
; SameSite settings. Either "none", "lax", or "strict"
SAME_SITE         = lax

[picture]
AVATAR_UPLOAD_PATH               = data/avatars
REPOSITORY_AVATAR_UPLOAD_PATH    = data/repo-avatars
; How Gitea deals with missing repository avatars
; none = no avatar will be displayed; random = random avatar will be displayed; image = default image will be used
REPOSITORY_AVATAR_FALLBACK       = none
REPOSITORY_AVATAR_FALLBACK_IMAGE = /img/repo_default.png
; Max Width and Height of uploaded avatars.
; This is to limit the amount of RAM used when resizing the image.
AVATAR_MAX_WIDTH                 = 4096
AVATAR_MAX_HEIGHT                = 3072
; Maximum alloved file size for uploaded avatars.
; This is to limit the amount of RAM used when resizing the image.
AVATAR_MAX_FILE_SIZE             = 1048576
; Chinese users can choose "duoshuo"
; or a custom avatar source, like: http://cn.gravatar.com/avatar/
GRAVATAR_SOURCE                  = gravatar
; This value will always be true in offline mode.
DISABLE_GRAVATAR                 = false
; Federated avatar lookup uses DNS to discover avatar associated
; with emails, see https://www.libravatar.org
; This value will always be false in offline mode or when Gravatar is disabled.
ENABLE_FEDERATED_AVATAR          = true

[attachment]
; Whether issue and pull request attachments are enabled. Defaults to `true`
ENABLED       = true
; Comma-separated list of allowed file extensions (`.zip`), mime types (`text/plain`) or wildcard type (`image/*`, `audio/*`, `video/*`). Empty value or `*/*` allows all types.
ALLOWED_TYPES = .docx,.gif,.gz,.jpeg,.jpg,.log,.pdf,.png,.pptx,.txt,.xlsx,.zip,.txt,.md,.xz,.rar,.tar,.bz2,.7z
; Storage type for attachments, `local` for local disk or `minio` for s3 compatible
; object storage service, default is `local`.
STORAGE_TYPE  = local
; Path for attachments. Defaults to `data/attachments` only available when STORAGE_TYPE is `local`
PATH          = data/attachments

[log]
ROOT_PATH           = /var/lib/gitea/log
; Either "console", "file", "conn", "smtp" or "database", default is "console"
; Use comma to separate multiple modes, e.g. "console, file"
MODE                = console
; Buffer length of the channel, keep it as it is if you don't know what it is.
BUFFER_LEN          = 10000
; Either "Trace", "Debug", "Info", "Warn", "Error", "Critical", default is "Info"
; TEMP Info => Debug to debug TFA
ROUTER_LOG_LEVEL    = Debug
ROUTER              = console
ENABLE_ACCESS_LOG   = false
ACCESS_LOG_TEMPLATE = {{.Ctx.RemoteAddr}} - {{.Identity}} {{.Start.Format "[02/Jan/2006:15:04:05 -0700]" }} "{{.Ctx.Req.Method}} {{.Ctx.Req.URL.RequestURI}} {{.Ctx.Req.Proto}}" {{.ResponseWriter.Status}} {{.ResponseWriter.Size}} "{{.Ctx.Req.Referer}}\" \"{{.Ctx.Req.UserAgent}}"
ACCESS              = file
; Either "Trace", "Debug", "Info", "Warn", "Error", "Critical", default is "Trace"
; TEMP Info => Debug to debug TFA
LEVEL               = Debug
; Either "Trace", "Debug", "Info", "Warn", "Error", "Critical", default is "None"
STACKTRACE_LEVEL    = None

; Generic log modes
[log.x]
FLAGS    = stdflags,utc
COLORIZE = false

; For "console" mode only
[log.console]
STDERR = false

[mirror]
; Default interval as a duration between each check
DEFAULT_INTERVAL = 8h
; Min interval as a duration must be > 1m
MIN_INTERVAL     = 10m

[oauth2]
; Enables OAuth2 provider
ENABLE                        = true
; Lifetime of an OAuth2 access token in seconds
ACCESS_TOKEN_EXPIRATION_TIME  = 3600
; Lifetime of an OAuth2 refresh token in hours
REFRESH_TOKEN_EXPIRATION_TIME = 730
; Check if refresh token got already used
INVALIDATE_REFRESH_TOKENS     = false
; OAuth2 authentication secret for access and refresh tokens, change this yourself to a unique string. CLI generate option is helpful in this case. https://docs.gitea.io/en-us/command-line/#generate
JWT_SECRET                    = xxxxxxxxxxxxxxxxxxxxxxxxxxxx
; Maximum length of oauth2 token/cookie stored on server
MAX_TOKEN_LENGTH              = 32767

[i18n]
LANGS = en-US,zh-CN,zh-HK,zh-TW,de-DE,fr-FR,nl-NL,lv-LV,ru-RU,uk-UA,ja-JP,es-ES,pt-BR,pt-PT,pl-PL,bg-BG,it-IT,fi-FI,tr-TR,cs-CZ,sr-SP,sv-SE,ko-KR
NAMES = English,简体中文,繁體中文（香港）,繁體中文（台灣）,Deutsch,français,Nederlands,latviešu,русский,Українська,日本語,español,português do Brasil,Português de Portugal,polski,български,italiano,suomi,Türkçe,čeština,српски,svenska,한국
어

[other]
SHOW_FOOTER_BRANDING           = false
; Show version information about Gitea and Go in the footer
SHOW_FOOTER_VERSION            = true
; Show template execution time in the footer
SHOW_FOOTER_TEMPLATE_LOAD_TIME = false

[task]
; Task queue type, could be `channel` or `redis`.
QUEUE_TYPE   = channel
; Task queue length, available only when `QUEUE_TYPE` is `channel`.
QUEUE_LENGTH = 1000

[migrations]
; Max attempts per http/https request on migrations.
MAX_ATTEMPTS  = 3
; Backoff time per http/https request retry (seconds)
RETRY_BACKOFF = 3

; default storage for attachments, lfs and avatars
[storage]
; storage type
STORAGE_TYPE = local

; lfs storage will override storage
[lfs]
STORAGE_TYPE = local
```

Note that I'm not sure in this app.ini file (which I did a quick cleanup of) what's left that is by default and thus can be removed and what isn't. I'm off to bed for now though, so I'll perhaps think about looking it up later.

Thanks.

**Comment by wxiaoguang** [comment]
> Then I went to set it up again, and using both a desktop TOTP generator and copying the secret into it, as well as using the QR code for a mobile TFA app and made both generate the 6-digit codes, and both generate the exact same codes and switch codes at the exact same time, however, the Enroll into TFA page on my gitea instance is claiming that the new passcodes generated by both using the secret it provides is invalid somehow.

A small question, have you checked your system time is accurate and correct? You can check your system time here: https://time.is

**Comment by Kusoneko** [comment]
> > Then I went to set it up again, and using both a desktop TOTP generator and copying the secret into it, as well as using the QR code for a mobile TFA app and made both generate the 6-digit codes, and both generate the exact same codes and switch codes at the exact same time, however, the Enroll into TFA page on my gitea instance is claiming that the new passcodes generated by both using the secret it provides is invalid somehow.
> 
> A small question, have you checked your system time (running Gitea instance) is accurate and correct? You can check your system time here: https://time.is . Or check time by google:
> 
> ```
> $ curl -v "https://google.com" 2>&1 | grep -i "date:" | grep -v "^\*"
> < date: Fri, 03 Sep 2021 06:57:17 GMT
> 
> $ date -u
> Fri Sep  3 06:57:17 UTC 2021
> ```

```
$ curl -v "https://google.com" 2>&1 | grep -i "date:" | grep -v "^\*"
< date: Fri, 03 Sep 2021 07:12:44 GMT
$ date -u
Fri Sep  3 07:13:55 AM UTC 2021
```

**Comment by wxiaoguang** [comment]
You see, it's your system time's problem.

TOTP won't work correctly if the error between your server's time and client's time is greater than a few seconds.

So it's not Gitea's bug 😊

**Comment by wxiaoguang** [comment]
PS: Maybe the TOTP problem should be written into FAQ. I know many people use servers with wrong system time.

**Comment by zeripath** [comment]
@wxiaoguang I meant to suggest timing but I was distracted by them saying it had previously worked but yes this would be consistent with that being the issue.

So I'm gonna close.

**Comment by Kusoneko** [comment]
Just to confirm, I went and enabled and started systemd-timesyncd that was for some reason not enabled nor started, and tried again and this time it worked. Thank you everyone!
